feat(storagezone): add the ability to add, get, list, and remove stor…

jspizziri opened the pull request:
…age zones

jspizziri commented:
My overarching goal is to [get storage zone management into the TF provider you've written](https://github.com/simplesurance/terraform-provider-bunny/issues/62). My Golang-foo is weak (this is in fact the first Go I've ever written. There's a lot of repetition from the `pullzone` code and I think a more generic `resource` that has abstractions for Add, Get, List, Pagination, etc would be best. I took a look at Go generics and played around with them for a few minutes, and couldn't immediately get them to work for me. I figured I'd push this up as something that works is more valuable than something that's better but ultimately doesn't exist.

Would be interested in your thoughts and guidance.

fho commented on storagezone_add.go:
```diff
@@ -0,0 +1,38 @@
+// Bunny.net API docs: https://docs.bunny.net/reference/storagezonepublic_add
+type StorageZoneAddOptions struct {
+	// The origin URL of the storage zone where the files are fetched from.
+	OriginURL string `json:"OriginUrl,omitempty"`
+	// The ID of the storage zone that the storage zone is linked to.
+	Region string `json:"Region,omitempty"`
+
+	// The name of the storage zone (Optional)
+	Name string `json:"Name,omitempty"`
+	// The code of the main storage zone region (Optional)
+	ReplicationRegions string `json:"ReplicationRegions,omitempty"`
```
Please use pointers for those fields.
The bunny api does not define if it always assumes the same values then Go if fields are omitted.
By using pointers we can distinguish clearly between unset and default values. This prevents discrepancy when interpreting the unset fields.

fho reviewed (commented):
@jspizziri thanks a lot for your contribution!
It looks great overall.

Please see my review comments + also check out the warnings from the CI check.

jspizziri commented:
@fho please see my updates. I believe I've taken care of everything you mentioned. I refactored things to use functions with generics and I had to upgrade to go `1.18` to do so. LMK what you think.

fho commented:
@jspizziri CI is failing because it uses Go 1.17 (needs to be upgraded in .github/workflows/ci.yml).

I would prefer if you could split those changes into multiple PRs or at least multiple commits.
Updating Go, supporting the storage API and refactoring the code to generics are separate topics.

It makes it easier to use the commit messages as base for changelogs, enables to revert specific changes that have introduced bugs, makes it easier to review PRs and it's also quicker to finish small changes and get them into master. :-)

jspizziri commented:
@fho please see my updates.

jspizziri commented:
@fho the workflow is failing due to the linter. I checked and [it looks like `golangci-lint` might not support go `1.18` yet](https://github.com/golangci/golangci-lint/issues/2649).

fho commented:
@jspizziri awesome, thanks a lot.

I like the idea with the generic resource functions!
I'll review it more thoroughly latest on Thursday, have to read a bit about go-generics again first, haven't used them so far. :-)

jspizziri commented:
FYI I'm adding `update` as well since it's required by TF. I'll let you know once it's there, but it will be almost identical to everything else in terms of design so please still take a look at the generics usage.

jspizziri commented:
Alright, I've updated the client to add `update` to storage zones. A few things of note:

1. I modified the `client` to check for a `204` status code and not require a response (204 is an empty response code).
2. There was a notable naming inconsistency with the `ReplicationRegions/ReplicationZones` property.

fho reviewed (commented):
looks great, I added some comments

jspizziri commented on storagezone_get.go:
```diff
@@ -0,0 +1,29 @@
+	FilesStored        *int64      `json:"FilesStored,omitempty"`
+	Region             *string     `json:"Region,omitempty"`
+	ReplicationRegions []string    `json:"ReplicationRegions,omitempty"`
+	PullZones          []*PullZone `json:"PullZones,omitempty"`
```
I can confirm that its the exact same object. In the process of doing the diff I noticed that there are some missing properties from the existing `PullZone` object (perhaps added since the object was created). I'll add those as well.

fho replied:
thanks a lot

Could you please create a separate PR to add the missing fields to the Pull-Zone? :-)
It's a different topic which we can process separately. 
I would like to keep the PR from getting bigger and bigger, it becomes more effort to review and 
 the actual functionality that you need (storage zone api) is getting delayed because of unrelated changes.

jspizziri commented on storagezone_add.go:
```diff
@@ -0,0 +1,32 @@
+	// The ID of the storage zone that the storage zone is linked to.
+	Region *string `json:"Region,omitempty"`
+
+	// The name of the storage zone (Optional)
```
You're correct, this is just a copy/paste error.

jspizziri commented on client.go:
```diff
@@ -301,7 +303,7 @@ func (c *Client) unmarshalHTTPJSONBody(resp *http.Response, reqURL string, resul
 	}
 
 	if len(body) == 0 {
-		if result != nil {
+		if result != nil && resp.StatusCode != 204 {
```
yeah, so I tried doing something like that initially, but AFAIK there is no such thing as an `undefined` type. I thought of creating a special struct like `EmptyResponse` or something that we would check for instead. What do you think about that?

jspizziri replied:
I implemented a `NoContentResponse` type, which I believe accomplishes what you want, may not be the most ideal way but the only one that was apparent with my limited Go skills. LMK what you think.

fho replied:
I experimented a bit around and found a way how we can forward nil.

Change the resourceAdd function to accept a pointer to the response, this will allow to pass nil:

```golang
func resourceAdd[Resp any](
	ctx context.Context,
	client *Client,
	path string,
	opts any,
	resp *Resp,
) (*Resp, error) {
	req, err := client.newPostRequest(path, opts)
	if err != nil {
		return nil, err
	}

	if err := client.sendRequest(ctx, req, resp); err != nil {
		return nil, err
	}

	return resp, nil
}
```

The functions that call resourceAdd can still kept short by using named return values, like:
```golang
func (s *PullZoneService) Add(ctx context.Context, opts *PullZoneAddOptions) (resp *PullZone, err error) {
	return resourceAdd(
		ctx,
		s.client,
		"/pullzone",
		opts,
		resp,
	)
}
```

I think it's less complex then introducing the NoContentResponse type + using reflection.

What do you think?

jspizziri replied:
It seems to me that this approach would actually require more effort in the long run, as every time you call a `resource*` function you'll need to pass an additional parameter. Whereas, with `NoContentResponse` you only need to opt into that on the specific methods that actual return a `204`. But I'm not a Go expert so I might be missing how your proposal actually works in practice. With that said:

1. Again, I'm a Go noob, so I defer your expertise.
2. You're the boss 😄

jspizziri replied:
My latest push implements my last concept. Take a look and if you still want to remove `NoContentResponse` lmk and I can do that.

jspizziri commented on resource_list.go:
```diff
@@ -0,0 +1,76 @@
+}
+
+// ResourceList is a generic function for listing resources
+func ResourceList[Resp any](
```
I've updated it to do something similar to what you're suggesting. Take a look and let me know what you think.

fho replied:
Good idea to invert the relationship. 

I also did a bit of research now and it seems not possible to only define that a type has to have certain struct field  embeds another structs.
Adding getter methods for PaginationReply + defining an interface which is used for the type constraint, should work.
The package user that probably can not access the PaginationReply fields directly and would have to call getter-method. I think that would be much worse too use.

I prefer the previous version, where `Resp` is any. 
With the new change the public API of the package also starts to use generics, which makes it more complex for the package user, only because we want to replace 2 functions calls with one internally.
I think that is not worth it.

jspizziri replied:
I changed it back to `any` and left the inverted relationship.

fho commented on storagezone_add.go:
```diff
@@ -0,0 +1,32 @@
+	Name *string `json:"Name,omitempty"`
+	// The code of the main storage zone region (Optional)
+	ReplicationRegions []string `json:"ReplicationRegions,omitempty"`
+}
```
I would not define an enum (aka in Golang an own type + consts with that type) in this case because it's possible that the enum will be incomplete. This would make it a bit weird to use.
I would only define strings consts like:
```go
const (
  StorageZoneRegionDE = "DE"
  StorageZoneRegionNY = "NY"
...
)
const (
  StorageZoneReplicationRegionDE = "DE"
...
)
```

fho commented:
@jspizziri there are multiple mehtods that do not use the new resource functions, e.g.: AddCustomCertificate, RemoveCustomHostname, SetEdgeRuleEnabled could you change all to use the new resource functions, so it is consistent?

fho commented:
@jspizziri fyi, I made https://github.com/simplesurance/bunny-cli public.
I created the tool to test the bunny-go package manually and make experimenting with the bunny-go api a bit easier.
Often using curl + jq works similar well and testing bunny-go is better done with testcases. :-)
Maybe it's still helpful to you

jspizziri commented:
I've pushed a new update. Some things of note:

1.  Refactored all the pullzone functions to use the abstracted `resource*` functions
2. In doing so, I had to abstract the `resource` functions a bit more (pulled path interpolation out and added some more options)
3. I removed `resource_add` and `resource_update` in favor of a `resource_post` which has functions for getting a response with & without content. Bunny doesn't really have a true `update/PUT` endpoint, so it reflects more closely their actual API.

LMK your thoughts from here

jspizziri commented:
Also, something I want to note that I just discovered about on PullZones:

The docs say that on creation of a PZ an OriginURL is required and a StorageZoneId is optional. However, the reality is that ONE of the TWO are required. You must either specify a OriginURL OR a StorageZoneID. Furthermore, passing BOTH results in OriginURL being entirely ignored (StorageZoneID is preferred).

I found this while updating the TF provider to support Storage Zones. I feel that the libs should help in this regard, and my thought was that `bunny-go` could log a warning if you pass both at the same time and that the TF provider could throw an error if you pass both at the same time (since it creates issues with the TF diff make it seem like there are changes to apply when there are not)

jspizziri commented:
Also, it should be noted that the refactor of pagination actually introduces a breaking change since it removes some old structs in favor of abstract ones.

fho commented:
> The docs say that on creation of a PZ an OriginURL is required and a StorageZoneId is optional. However, the reality is that ONE of the TWO are required. You must either specify a OriginURL OR a StorageZoneID. Furthermore, passing BOTH results in OriginURL being entirely ignored (StorageZoneID is preferred).

nice find, lets document it in the godoc for `PullZoneAddOptions` and log a message if both are passed as you suggested in a separate PR

fho commented:
@jspizziri looks good! thanks a for the work again.

I'm in favor of removing `resourceUpdate204` and pass `resp` to `resourcePost` explicitly. I think it's more simple and clear.
I also can do that before I merge the PR.

My plan is to fix the failing integrationtest in the bunny-terraform-go provider, then run the the terraform provider with this bunny-go version to test the change thoroughly. I expect that everything works fine but better be safe. :-)
Afterwards I'll merge it.

jspizziri commented:
> > The docs say that on creation of a PZ an OriginURL is required and a StorageZoneId is optional. However, the reality is that ONE of the TWO are required. You must either specify a OriginURL OR a StorageZoneID. Furthermore, passing BOTH results in OriginURL being entirely ignored (StorageZoneID is preferred).
> 
> nice find, lets document it in the godoc for `PullZoneAddOptions` and log a message if both are passed as you suggested in a separate PR

https://github.com/simplesurance/bunny-go/issues/24

jspizziri commented:
@fho thanks! Thanks for all the work on this lib and the tf provider. Definitely going to be super helpful for us.

fho commented:
@jspizziri yes, they are aware of both. I also told them that it would be great if they could list them on their 3. party tools page but it did not happen :shrug:.
